### PR TITLE
Add `isFieldEmpty` helper

### PR DESCRIPTION
**Pull request by xrav3nz**
This will be used as part of an form field's dirtiness check.

This is part of #22534.

/cc @cvializ @GoTcWang

### CODE REVIEW

File: src/form.js [hunk 1]
```diff
@@ -187,3 +187,24 @@ export function isFieldDefault(field) {
...
+export function isFieldEmpty(field) {
+  switch (field.type) {
+    case 'select-multiple':
+    case 'select-one':
+      // dropdown menu has at least one option always selected
+      return false;
+    case 'checkbox':
+    case 'radio':
+      return !field.checked;
+    default:
```
Comment (GoTcWang) [inline src/form.js]: Let's do an exhaustive switch and throw an error by default. (and also a test for the default case)

File: test/unit/test-form.js [hunk 2]
```diff
@@ -585,3 +590,101 @@ describes.realWin('isFieldDefault', {}, env => {
...
+    it('always returns false because at least one option is always selected', () => {
+      const dropdown = doc.createElement('select');
+
+      const optionA = doc.createElement('option');
+      dropdown.appendChild(optionA);
+      const optionB = doc.createElement('option');
+      dropdown.appendChild(optionB);
+
+      optionA.selected = false;
+      expect(optionA.selected).to.be.true;
```
Comment (GoTcWang) [inline test/unit/test-form.js]: Why are we asserting this?
This test doesn't need to test the behavior of <select> and <option> - we don't care if optionA or optionB is selected. We only need to test isFieldEmpty always returns false.

So I think what you can do is:
optionA.selected = false;
optionB.selected = false;

expect(isFieldEmpty(dropdown)).to.be.false;

Comment (GoTcWang) [inline test/unit/test-form.js]: nit: Add a new line between Arrange and Act block, this test is not a two line test anymore.
In this test you put Act and Assert block together, which is fine. But remember to add another new line between them if you separate them up, here or in other tests.

Comment (xrav3nz) [reply test/unit/test-form.js]: The original intention was to show that the option is always selected, even after being set to false. Will make the changes!

Comment (GoTcWang) [reply test/unit/test-form.js]: Yeah that's good. 
If you tend to test that, you can have another test with a proper description. This test, as the name suggests, should focus on if isEmptyField returns false.

If you are going to write that test, I still feel you should set both option to be not selected and assert on one of them being selected anyway. That explains the problem clearly and make optionA and optionB order independent.

File: src/form.js (hunk 1 above)
Comment (xrav3nz) [reply src/form.js]: Do you think an enumeration of all `type`s is necessary? Or could we get away with checking if it's an `input` element? (There are ~20 `type`s valid in AMP right now)

Comment (GoTcWang) [reply src/form.js]: Okay now I understand why they don't have an exhaustive switch in the other helper......

However, as you point out, I am still confused why we don't check elements first. Both this and `isFieldDefault` helpers don't make sense on elements other than <input> <select> and <amp-selector> and etc., but the call site didn't specify.

I suggest we switch on the element tag names:

```js
switch(element.tagName):
  case input:
     if filed.type === radio or field.type === radio
        return ...;
     else
        return ...;
  case select:
     return false;
  case amp-selector:
    //TODO
  default:
     return false;
```

But I might be missing something, in that case, we should at least document in the JsDoc what kind of elements which be called with this helper.

Comment (GoTcWang) [reply src/form.js]: Not sure if it's a bug or never supported, I can't put code snippets with Md here. Anyway, code snippets:
https://jsfiddle.net/698aqzLx/

File: src/form.js [hunk 3]
```diff
@@ -187,3 +187,27 @@ export function isFieldDefault(field) {
...
+      } else {
+        return !field.value;
+      }
+    case 'TEXTAREA':
+      return !field.value;
+    case 'SELECT':
+      // dropdown menu has at least one option always selected
+      return false;
+    default:
+      throw new Error(`${field.tagName} is not a supported field element.`);
```
Comment (xrav3nz) [inline src/form.js]: @GoTcWang Changed the conditional to `switch` on `tagName`, and also throws an error upon unsupported field element. I used this existing error message as a reference:

https://github.com/ampproject/amphtml/blob/95c5ac41e773e54630ac64e9e9ff53259a18c6c7/extensions/amp-bind/0.1/bind-expression.js#L374

Comment (GoTcWang) [reply src/form.js]: Discussed offline. This helper is only called for those elements triggering `input` event so throwing an error is the correct way to go, I think. However, we should definitely document this in the JsDoc so it won't be misused later.

File: test/unit/test-form.js [hunk 4]
```diff
@@ -585,3 +590,110 @@ describes.realWin('isFieldDefault', {}, env => {
...
 });
+
+describes.fakeWin('isFieldEmpty', {}, env => {
+  let doc;
+
+  beforeEach(() => {
+    doc = env.win.document;
+  });
+
+  describe('text field', () => {
```
Comment (GoTcWang) [inline test/unit/test-form.js]: nit: rearrange tests in the order similar to your implementation. 

- "input element" -> "type checkbox" -> empty/not empty
- "input element" -> "type radio" -> empty/not empty
- "input element" -> "type others" -> empty/not empty
- "textarea" -> empty/not empty
- "select"
-  "default"

File: src/form.js (hunk 3 above)
Comment (GoTcWang) [inline src/form.js]: `${field.tagName} is not a supported field element to check emptiness?

Comment (xrav3nz) [reply src/form.js]: I appended to the method name to the message `isFieldEmpty: tag is not a supported field element`

Comment (xrav3nz) [reply src/form.js]: I explicitly stated the supported elements in the JsDoc, but deliberately left out the `input event` implementation detail.

**Pull request merged**